### PR TITLE
fix: Make Push and Pull Update Cache to Work with VSCE Conflict Detection

### PR DESCRIPTION
**Pull request by klewis-sfdc**
### What does this PR do?
* Updates the local cache with timestamps after push and pull operations in order for these commands to work correctly when used with Deploy and the "detect conflicts on sync" setting.

### What issues does this PR fix or reference?
#4848, @W-13190834@

### Functionality Before
* when used together in a source tracked org (push/pull+deploy), local changes could be reported as conflicts

### Functionality After
* push pull and deploy can be used interchangeably with "conflict detection at sync" enabled and it still allows local changes to be deployed when there have not been more recent changes on the server.

**Notes:**
The only remaining command extending `BaseDeployExecutor` was `ForceSourcePushExecutor`.  The logic that remained in `BaseDeployExecutor` was relocated and the baseDeployCommand.ts file was removed, simplifying the structure for these two commands:

_Before:_

_After:_

### CODE REVIEW

File: packages/salesforcedx-vscode-core/src/conflict/persistentStorageService.ts [hunk 1]
```diff
@@ -39,6 +40,38 @@ export class PersistentStorageService {
...
+  private bolsterChangedFiles(changedFiles: FileInfo[]): FileInfo[] {
+    // build a new array that adds '*-meta.xml' files for each .cls or .cmp file
+    const bolsteredChangedFiles = [];
+    for (const file of changedFiles) {
+      if (file.filePath) {
+        bolsteredChangedFiles.push(file);
+        const filePath = file.filePath;
+        const filePathLength = filePath.length;
+        const fileExtension = filePath.substring(filePathLength - 4);
+        if (fileExtension === '.cls' || fileExtension === '.cmp') {
```
Comment (klewis-sfdc) [inline packages/salesforcedx-vscode-core/src/conflict/persistentStorageService.ts:64]: STL only gives the main file in the response to the call to get status.  But we need to update the meta files in the cache also.  Not sure if other types need this too...

File: packages/salesforcedx-utils-vscode/src/services/types/FileInfo.ts [hunk 2]
```diff
@@ -0,0 +1,8 @@
+/*
+ * Copyright (c) 2023, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+export type FileInfo = { type: string; fullName: string; filePath?: string };
```
Comment (gbockus-sf) [inline packages/salesforcedx-utils-vscode/src/services/types/FileInfo.ts:8]: what is type here?

Comment (klewis-sfdc) [reply packages/salesforcedx-utils-vscode/src/services/types/FileInfo.ts:8]: I just created it.  I wasn't sure what to do with the types here.  I get StatusOutputRow[] back from STL.  I guess I can pass that straight to PersistentStorageService...

Comment (klewis-sfdc) [reply packages/salesforcedx-utils-vscode/src/services/types/FileInfo.ts:8]: Oh wait, maybe I'll export the StatusOutputRow type from SourceTrackingService...

File: packages/salesforcedx-vscode-core/src/commands/util/sfdxCommandlet.ts [hunk 3]
```diff
@@ -126,6 +139,8 @@ export abstract class SfdxCommandletExecutor<T>
   }
 
   public abstract build(data: T): Command;
+  protected getRemoteChanges?(): StatusOutputRowType[] | undefined;
+  protected updateCacheAfterPushPull?(): void;
```
Comment (gbockus-sf) [inline packages/salesforcedx-vscode-core/src/commands/util/sfdxCommandlet.ts:143]: nitty suggestion, you could have this method be called updateCache and just have it default to a noop function, then override it in the push/pull case.  Then you wouldn't have the sometimes defined method you'd just alway execute it.

File: packages/salesforcedx-vscode-core/src/commands/util/sfdxCommandlet.ts [hunk 4]
```diff
@@ -52,7 +63,10 @@ export abstract class SfdxCommandletExecutor<T>
     cancellationTokenSource: vscode.CancellationTokenSource,
     cancellationToken: vscode.CancellationToken
   ) {
-    channelService.streamCommandOutput(execution);
+    const p = execution.command.logName;
+    if (p !== FORCE_SOURCE_PULL_LOG_NAME) {
+      channelService.streamCommandOutput(execution);
+    }
```
Comment (klewis-sfdc) [inline packages/salesforcedx-vscode-core/src/commands/util/sfdxCommandlet.ts:69]: I added this to suppress the JSON output from being shown in the output panel

File: packages/salesforcedx-vscode-core/src/commands/util/sfdxCommandlet.ts [hunk 5]
```diff
@@ -95,28 +111,55 @@ export abstract class SfdxCommandletExecutor<T>
...
+      exitCode === 0 &&
+      execution.command.logName === FORCE_SOURCE_PULL_LOG_NAME
+    ) {
+      const pullResult = JSON.parse(output);
+      this.updateCache(pullResult);
+
+      const pullParser = new ForcePullResultParser(output);
+      const errors = pullParser.getErrors();
+      if (errors) {
+        channelService.showChannelOutput();
```
Comment (klewis-sfdc) [inline packages/salesforcedx-vscode-core/src/commands/util/sfdxCommandlet.ts:137]: Not sure about this part where errors are shown if encountered - need to generate or fake an error in the pull op to see how this looks.

File: packages/salesforcedx-utils-vscode/src/cli/pullResultParser.ts [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+export class ForcePullResultParser {
+  private response: any;
+
+  constructor(stdout: string) {
+    try {
+      this.response = extractJsonObject(stdout);
+    } catch (e) {
+      const err = new Error('Error parsing pull result');
```
Comment (gbockus-sf) [inline packages/salesforcedx-utils-vscode/src/cli/pullResultParser.ts:43]: Does this get displayed to the user (should it be externalized?)

File: packages/salesforcedx-vscode-core/src/commands/util/sfdxCommandlet.ts [hunk 7]
```diff
@@ -52,7 +63,12 @@ export abstract class SfdxCommandletExecutor<T>
     cancellationTokenSource: vscode.CancellationTokenSource,
     cancellationToken: vscode.CancellationToken
   ) {
-    channelService.streamCommandOutput(execution);
+    const commandLogName = execution.command.logName;
+    // If Pull operation, output text will be
+    // generated later by ForcePullResultParser
```
Comment (gbockus-sf) [inline packages/salesforcedx-vscode-core/src/commands/util/sfdxCommandlet.ts:68]: nit: p -> logName

File: packages/salesforcedx-vscode-core/src/commands/util/sfdxCommandlet.ts (hunk 5 above)
Comment (gbockus-sf) [reply packages/salesforcedx-vscode-core/src/commands/util/sfdxCommandlet.ts:137]: jest test FTW

File: packages/salesforcedx-vscode-core/src/commands/util/sfdxCommandlet.ts [hunk 8]
```diff
@@ -95,28 +111,55 @@ export abstract class SfdxCommandletExecutor<T>
...
+    execution: CliCommandExecution,
+    response: ContinueResponse<T>,
+    startTime: [number, number],
+    output: string
+  ): void {
+    if (
+      exitCode === 0 &&
+      execution.command.logName === FORCE_SOURCE_PULL_LOG_NAME
+    ) {
+      const pullResult = JSON.parse(output);
```
[the remark this answers is not in the document]
Comment (gbockus-sf) [reply packages/salesforcedx-vscode-core/src/commands/util/sfdxCommandlet.ts:131]: def another good unit test  scenario

File: packages/salesforcedx-vscode-core/src/commands/util/sfdxCommandlet.ts [hunk 9]
```diff
@@ -125,6 +168,77 @@ export abstract class SfdxCommandletExecutor<T>
...
 
+  public outputResultPull(parser: ForcePullResultParser) {
+    const table = new Table();
+    const titleType = 'pull';
+
+    const successes: any = parser.getSuccesses();
+    const errors = parser.getErrors();
+    const pulledSource = successes ? successes.result.pulledSource : undefined;
+    if (pulledSource) {
+      const rows = pulledSource || (errors && errors.result);
```
Comment (gbockus-sf) [inline packages/salesforcedx-vscode-core/src/commands/util/sfdxCommandlet.ts:179]: nit: `errors?.result` can replace `errors && erros.result`

File: packages/salesforcedx-vscode-core/src/commands/util/sfdxCommandlet.ts [hunk 10]
```diff
@@ -125,6 +168,77 @@ export abstract class SfdxCommandletExecutor<T>
...
+      const tableTitle = nls.localize(`table_title_${titleType}ed_source`);
+      const outputTable = this.getOutputTable(table, rows, tableTitle);
+
+      channelService.appendLine(outputTable);
+      if (pulledSource && pulledSource.length === 0) {
```
Comment (gbockus-sf) [inline packages/salesforcedx-vscode-core/src/commands/util/sfdxCommandlet.ts:184]: In this flow does it show an empty table and then the empty message?

File: packages/salesforcedx-vscode-core/src/commands/util/sfdxCommandlet.ts [hunk 11]
```diff
@@ -125,6 +168,77 @@ export abstract class SfdxCommandletExecutor<T>
...
+    }
+
+    if (errors) {
+      const { name, message, result } = errors;
+      if (result) {
+        const outputTable = this.getErrorTable(table, result, titleType);
+        channelService.appendLine(outputTable);
+      } else if (name && message) {
+        channelService.appendLine(`${name}: ${message}\n`);
+      }
```
Comment (gbockus-sf) [inline packages/salesforcedx-vscode-core/src/commands/util/sfdxCommandlet.ts:197]: nit: do we need an else here to say something if there's no name & message, or result?

File: packages/salesforcedx-vscode-core/src/commands/util/sfdxCommandlet.ts [hunk 12]
```diff
@@ -125,6 +168,77 @@ export abstract class SfdxCommandletExecutor<T>
...
+    }
+  }
+
+  protected getOutputTable(
+    table: Table,
+    rows: unknown,
+    outputTableTitle: string | undefined
+  ) {
+    const outputTable = table.createTable(
+      (rows as unknown) as Row[],
```
Comment (gbockus-sf) [inline packages/salesforcedx-vscode-core/src/commands/util/sfdxCommandlet.ts:207]: weird. Is there a reason that `Row[]` isn't just the type of the parameter?

File: packages/salesforcedx-vscode-core/src/commands/util/sfdxCommandlet.ts [hunk 13]
```diff
@@ -125,6 +168,77 @@ export abstract class SfdxCommandletExecutor<T>
...
+
+  /**
+   * @description Base method (no-op) that is overridden by sub-classes
+   * forceSourcePush and forceSourcePull to update the local cache's
+   * timestamps post-operation, in order to be in sync for the
+   * "Detect Conflicts at Sync" setting.
+   */
+  protected updateCache(result: any): void {}
+
   public abstract build(data: T): Command;
```
Comment (gbockus-sf) [inline packages/salesforcedx-vscode-core/src/commands/util/sfdxCommandlet.ts:242]: I love me some method docs.  I dunno if you need the @description annotation though

Comment (gbockus-sf) [reply packages/salesforcedx-vscode-core/src/commands/util/sfdxCommandlet.ts:242]: Screenshot 2023-05-19 at 12 09 08 PM
vs
Screenshot 2023-05-19 at 12 09 19 PM

File: packages/salesforcedx-vscode-core/src/commands/baseDeployCommand.ts [hunk 14]
```diff
@@ -55,44 +54,70 @@ export abstract class BaseDeployExecutor extends SfdxCommandletExecutor<
...
+    stdOut: string,
+    workspacePath: string,
+    execFilePathOrPaths: string,
+    execution: CliCommandExecution,
+    startTime: [number, number],
+    cancellationToken: vscode.CancellationToken | undefined,
+    cancellationTokenSource: vscode.CancellationTokenSource
+  ): Promise<void> {
+    if (exitCode === 0 && this.getDeployType() === DeployType.Push) {
+      const pushResult = JSON.parse(stdOut);
```
Comment (gbockus-sf) [inline packages/salesforcedx-vscode-core/src/commands/baseDeployCommand.ts:81]: same q here about the JSON.parse.  what happen if the parse throws?

File: packages/salesforcedx-vscode-core/src/commands/baseDeployCommand.ts [hunk 15]
```diff
@@ -55,44 +54,70 @@ export abstract class BaseDeployExecutor extends SfdxCommandletExecutor<
...
-        BaseDeployExecutor.errorCollection.clear();
-        if (e.name !== 'DeployParserFail') {
-          e.message =
-            'Error while creating diagnostics for vscode problem view.';
+        } else {
+          success = true;
         }
-        telemetryService.sendException(e.name, e.message);
-        console.error(e.message);
+        this.outputResult(deployParser);
```
Comment (gbockus-sf) [inline packages/salesforcedx-vscode-core/src/commands/baseDeployCommand.ts:105]: what happens here if we get to the `handleDiagnosticErrors` call above?

File: packages/salesforcedx-vscode-core/src/commands/baseDeployCommand.ts [hunk 16]
```diff
@@ -55,44 +54,70 @@ export abstract class BaseDeployExecutor extends SfdxCommandletExecutor<
...
-      await DeployQueue.get().unlock();
-    });
-
+    } catch (e) {
+      BaseDeployExecutor.errorCollection.clear();
+      if (e.name !== 'DeployParserFail') {
+        e.message = 'Error while creating diagnostics for vscode problem view.';
+      }
+      telemetryService.sendException(e.name, e.message);
+      console.error(e.message);
```
Comment (gbockus-sf) [inline packages/salesforcedx-vscode-core/src/commands/baseDeployCommand.ts:113]: I know you are just moving this code, but it weird we don't tell the user anything if it goes sideways outside of the dev console.

File: packages/salesforcedx-vscode-core/src/conflict/persistentStorageService.ts [hunk 17]
```diff
@@ -75,6 +75,15 @@ export class PersistentStorageService {
     }
   }
 
+  public setPropertiesForFilesPushPull(files: any) {
```
Comment (gbockus-sf) [inline packages/salesforcedx-vscode-core/src/conflict/persistentStorageService.ts:78]: bummer we can't use a type here

**Review by gbockus-sf — COMMENTED** [review]
Some mostly nitty suggestions.  Great progress.  Thanks for figuring this out.

File: packages/salesforcedx-vscode-core/src/commands/util/sfdxCommandlet.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (klewis-sfdc) [reply packages/salesforcedx-vscode-core/src/commands/util/sfdxCommandlet.ts:131]: It looks like this
Error goes to the console
Toast notification just keeps spinning

File: packages/salesforcedx-vscode-core/src/commands/baseDeployCommand.ts (hunk 14 above)
Comment (klewis-sfdc) [reply packages/salesforcedx-vscode-core/src/commands/baseDeployCommand.ts:81]: Here it outputs to the log and ends with a 0 exit code

File: packages/salesforcedx-vscode-core/src/commands/util/sfdxCommandlet.ts [hunk 18]
```diff
@@ -95,28 +111,74 @@ export abstract class SfdxCommandletExecutor<T>
...
+      parsed = JSON.parse(output);
+    } catch (error) {
+      notificationService.showWarningMessage(
+        nls.localize('lib_retrieve_result_parse_error'),
+        output
+      );
+      throw error;
+    }
+    return parsed;
+  }
```
Comment (klewis-sfdc) [inline packages/salesforcedx-vscode-core/src/commands/util/sfdxCommandlet.ts:180]: This method now makes it look like this if there is a parsing error - let me know what you think

File: packages/salesforcedx-vscode-core/src/commands/util/sfdxCommandlet.ts (hunk 10 above)
Comment (klewis-sfdc) [reply packages/salesforcedx-vscode-core/src/commands/util/sfdxCommandlet.ts:203]: Nope, here it looks like this:

File: packages/salesforcedx-vscode-core/src/commands/util/sfdxCommandlet.ts (hunk 12 above)
Comment (klewis-sfdc) [reply packages/salesforcedx-vscode-core/src/commands/util/sfdxCommandlet.ts:226]: When I try I get this:

File: packages/salesforcedx-vscode-core/src/commands/baseDeployCommand.ts (hunk 15 above)
Comment (klewis-sfdc) [reply packages/salesforcedx-vscode-core/src/commands/baseDeployCommand.ts:105]: It looks like handleDiagnosticErrors is called when there are errors/conflicts, and it is supposed to output a table of the conflicts?  But right now it looks like it is just exiting and not showing a table at all:

To set this up, I created a conflict on the server for one file, then changed another one locally so I had two files to push, one with a conflict:

File: packages/salesforcedx-vscode-core/src/commands/util/sfdxCommandlet.ts [hunk 19]
```diff
@@ -95,28 +112,75 @@ export abstract class SfdxCommandletExecutor<T>
...
+   * @returns parsed JSON object.
+   */
+  protected parseOutput(output: string) {
+    let parsed: JSON;
+    try {
+      parsed = JSON.parse(output);
+    } catch (error) {
+      notificationService.showWarningMessage(
+        nls.localize('lib_retrieve_result_parse_error'),
+        output
```
Comment (gbockus-sf) [inline packages/salesforcedx-vscode-core/src/commands/util/sfdxCommandlet.ts:177]: nice.  Def way better.  I might just console.log the output so we can see if it this happens in production.  The 2nd parameter does indeed create buttons in the Warning modal.

File: packages/salesforcedx-vscode-core/src/commands/util/testData.ts [hunk 20]
```diff
@@ -0,0 +1,9 @@
+/*
+ * Copyright (c) 2023, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+export const dummyOutputPull =
```
Comment (gbockus-sf) [inline packages/salesforcedx-vscode-core/src/commands/util/testData.ts:8]: This is temporary right?

File: packages/salesforcedx-vscode-core/src/commands/baseDeployCommand.ts (hunk 16 above)
Comment (gbockus-sf) [reply packages/salesforcedx-vscode-core/src/commands/baseDeployCommand.ts:111]: should this return or throw?  Looks like with the current flow we're sending telemetry for both an exception and success.

File: packages/salesforcedx-vscode-core/src/conflict/persistentStorageService.ts [hunk 21]
```diff
@@ -4,7 +4,11 @@
  * Licensed under the BSD 3-Clause license.
  * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
  */
-import { getRootWorkspacePath } from '@salesforce/salesforcedx-utils-vscode';
+import {
+  DeployResult as DeployResultParsed,
+  getRootWorkspacePath
+} from '@salesforce/salesforcedx-utils-vscode';
+import { PullResult } from '@salesforce/salesforcedx-utils-vscode/src/cli/pullResultParser';
```
Comment (gbockus-sf) [inline packages/salesforcedx-vscode-core/src/conflict/persistentStorageService.ts:11]: Fix this import

**Comment by klewis-sfdc** [comment]
Unit tests are passing locally

**Review by gbockus-sf — APPROVED** [review]
QE Complete!
Looking great

Push no changes✅
Push With conflicts ✅
Push with conflicts override✅
Push with errors and conflicts✅
Pull no changes✅
Pull With conflicts✅
Pull with conflicts override✅
Pull with local error override✅
Deploy source from context (sanity check) ✅

Couple of questions to address in the QE docs & a couple of unit tests to fix and I think we're good to go.

I'll approve and feel free tom merge after those are addressed.

**Pull request merged**